main.py

```diff
@@ -1,18 +1,135 @@
+import numpy as np
+from scipy.stats import gaussian_kde
+import matplotlib.pyplot as plt
+
+TRAIN = True
+# training constants
+batch_size = 32
+epochs = 2
+train_fraction = 0.99
+num_priors = 100000
+num_sims_per_prior_pair = 1
+learning_rate = 0.0002
+num_train_samples = int(num_priors * num_sims_per_prior_pair * train_fraction)
+
 
 device = torch.device("cuda" if torch.cuda.is_available() else 'cpu')
 if torch.cuda.is_available():
     torch.set_default_tensor_type('torch.cuda.FloatTensor')
 
-model = Ratio(1, 1, 50)
+model = Ratio(1, 1, 200)
 model.to(device)
 
-def prior():
-    return torch.rand(1).to(device)
+prior = lambda: torch.rand(1).to(device)
+sim = DummySimulator()
+optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
+model.load_state_dict(torch.load("model.pt"))
+if TRAIN:
+    # %% GENERATE DATA
+    # generate pairs of priors
+    priors = [(prior(), prior()) for _ in range(num_priors)]
 
+    # generate data
+    print("Generating data...")
+    runs = []
+    for ts in tqdm(priors):
+        for _ in range(num_sims_per_prior_pair):
+            k = 0 if random() < 0.5 else 1
+            runs.append((k, sim.simulate(ts[k]), ts))
 
-sim = DummySimulator()
-ROLR(sim, model, prior, 500, 500000, 0.001)
+
+    # generate ratios
+    print("Generating ratios...")
+    ratios = [sim.eval_ratio(zs, ts[k], ts[(k + 1) % 2]) for k, zs, ts in tqdm(runs)]
+
+    # collate (add labels), shuffle, construct dataloaders
+    data = [(k, zs, ratio, ts) for ((k, zs, ts), ratio) in zip(runs, ratios)]
+    shuffle(data)
+    train_data, test_data = data[:num_train_samples], data[num_train_samples:]
+    train_loader = DataLoader(train_data, batch_size=batch_size, shuffle=True)
+    test_loader = DataLoader(test_data, batch_size=batch_size)
+
+    # %% TRAIN
+    print("Training...")
+    for i in range(epochs):
+        print("Epoch {}...".format(i))
+        train_iter = tqdm(train_loader)
+        model.train()
+        epoch_loss, num_samples = 0.0, 0
+        for batch in train_iter:
+            model.zero_grad()
+            labels, runs, ratios, (theta0, theta1) = batch
+            xs = runs[:, -1]
+            batch_sz = xs.shape[0]
+            r_hat = model(xs, theta0, theta1)
+            loss = rolr(r_hat, ratios, labels) * batch_size
+            loss.backward()
+            optimizer.step()
+            num_samples += len(labels)
+            epoch_loss += loss.item()
+            train_iter.set_description("Average loss {}".format(epoch_loss/num_samples))
+
+    # %% TEST
+    print("Testing...")
+    epoch_loss, num_samples = 0.0, 0
+    x, ratio_true, ratio_pred = np.zeros((0)), np.zeros((0)), np.zeros((0))
+    with torch.no_grad(): # we can only use this when we don't need the score (only the ratio)
+        test_perf, num_samples = 0.0, 0
+        test_iter = tqdm(test_loader)
+        for batch in test_iter:
+            labels, runs, ratios, (theta0, theta1) = batch
+            xs = runs[:, -1]
```

Comment:
If we change this to xs = runs[:, [-1]] we don't need to squeeze later in ratio.py

Reply:
Good point